File: src/java/WebVulnerabilities.java
package java;

import javax.servlet.http.*;
import java.io.*;

/**
 * This class contains web vulnerabilities (XSS, etc.).
 * Used for testing SAST tools.
 */
public class WebVulnerabilities extends HttpServlet {

    /**
     * Reflected XSS vulnerability
     */
    protected void doGet(HttpServletRequest req, HttpServletResponse resp)
            throws IOException {
        // VULNERABLE: Direct output of user input
        String name = req.getParameter("name");
        resp.getWriter().println("<h1>Hello " + name + "</h1>");
    }

    /**
     * Stored XSS vulnerability simulation
     */
    public String displayComment(String userComment) {
        // VULNERABLE: No output encoding
        return "<div class='comment'>" + userComment + "</div>";
    }

    /**
     * HTTP Response Splitting vulnerability
     */
    protected void doPost(HttpServletRequest req, HttpServletResponse resp)
            throws IOException {
        // VULNERABLE: User input in HTTP headers
        String location = req.getParameter("redirect");
        resp.setHeader("Location", location);
        resp.sendRedirect(location);
    }

    /**
     * Open Redirect vulnerability
     */
    public void redirectUser(HttpServletResponse resp, String url)
            throws IOException {
        // VULNERABLE: No validation of redirect URL
        resp.sendRedirect(url);
    }

    /**
     * Insecure cookie handling
     */
    protected void setCookies(HttpServletResponse resp, String userId) {
        // VULNERABLE: Cookie without HttpOnly flag
        Cookie cookie = new Cookie("sessionId", userId);
        cookie.setMaxAge(3600);
        // VULNERABLE: Missing secure flag in production
        resp.addCookie(cookie);
    }

    /**
     * Safe XSS prevention with output encoding
     */
    public String safeDisplayComment(String userComment) {
        // SAFE: Encode HTML special characters
        String encoded = userComment
            .replace("&", "&amp;")
            .replace("<", "&lt;")
            .replace(">", "&gt;")
            .replace("\"", "&quot;")
            .replace("'", "&#39;");
        return "<div class='comment'>" + encoded + "</div>";
    }
}

File: src/javascript/web_vulnerabilities.js
/**
 * Web vulnerabilities for JavaScript (Node.js and browser).
 * Used for testing SAST tools.
 */

const express = require('express');
const app = express();

/**
 * Reflected XSS vulnerability (Node.js/Express)
 */
app.get('/xss', (req, res) => {
    const name = req.query.name || '';

    // VULNERABLE: Direct output of user input
    res.send(`<h1>Hello ${name}</h1>`);
});

/**
 * Stored XSS simulation
 */
function displayComment(comment) {
    // VULNERABLE: No output encoding
    return `<div class="comment">${comment}</div>`;
}

/**
 * DOM-based XSS
 */
function domXSS() {
    // Browser-side vulnerable code
    const userInput = window.location.hash.substring(1);

    // VULNERABLE: innerHTML with user input
    document.getElementById('output').innerHTML = userInput;
}

/**
 * Open Redirect vulnerability
 */
app.get('/redirect', (req, res) => {
    const url = req.query.url || '/';

    // VULNERABLE: No validation of redirect URL
    res.redirect(url);
});

/**
 * HTTP Response Splitting
 */
app.get('/response-splitting', (req, res) => {
    const userAgent = req.headers['user-agent'] || '';

    // VULNERABLE: User input in headers
    res.set('X-User-Agent', userAgent);
    res.send('OK');
});

/**
 * Insecure cookie handling
 */
app.get('/set-cookie', (req, res) => {
    const userId = req.query.userId || '';

    // VULNERABLE: Missing HttpOnly and Secure flags
    res.cookie('sessionId', userId, { maxAge: 3600000 });
    res.send('Cookie set');
});

/**
 * Server-Side Request Forgery (SSRF)
 */
app.get('/ssrf', async (req, res) => {
    const url = req.query.url || '';
    const https = require('https');

    // VULNERABLE: User controls URL
    https.get(url, (response) => {
        let data = '';
        response.on('data', (chunk) => {
            data += chunk;
        });
        response.on('end', () => {
            res.send(data);
        });
    }).on('error', (err) => {
        res.status(500).send('Error');
    });
});

/**
 * JSON Injection
 */
app.get('/json-injection', (req, res) => {
    const userData = req.query.data || '{}';

    // VULNERABLE: Direct JSON parsing without validation
    const data = JSON.parse(userData);

    // VULNERABLE: JSONP callback injection
    const callback = req.query.callback || 'callback';
    res.set('Content-Type', 'application/javascript');
    res.send(`${callback}(${JSON.stringify(data)})`);
});

/**
 * Prototype Pollution
 */
function mergeObjects(target, source) {
    // VULNERABLE: Prototype pollution
    for (const key in source) {
        if (source.hasOwnProperty(key)) {
            target[key] = source[key];
        }
    }
    return target;
}

/**
 * Regex Denial of Service (ReDoS)
 */
function vulnerableRegex(input) {
    // VULNERABLE: Exponential time regex
    const regex = /^([a-zA-Z0-9]+)*$/;
    return regex.test(input);
}

/**
 * Unsafe deserialization
 */
function unsafeDeserialization(serialized) {
    // VULNERABLE: Using eval for deserialization
    return eval(`(${serialized})`);
}

// Safe examples
/**
 * Safe XSS prevention
 */
app.get('/safe-xss', (req, res) => {
    const name = req.query.name || '';

    // SAFE: Using template engine with auto-escaping
    // Or manually escape
    const escapedName = escapeHtml(name);
    res.send(`<h1>Hello ${escapedName}</h1>`);
});

function escapeHtml(text) {
    const map = {
        '&': '&amp;',
        '<': '&lt;',
        '>': '&gt;',
        '"': '&quot;',
        "'": '&#039;'
    };
    return text.replace(/[&<>"']/g, (m) => map[m]);
}

/**
 * Safe redirect with validation
 */
app.get('/safe-redirect', (req, res) => {
    const url = req.query.url || '/';

    // SAFE: Validate redirect URL
    const allowedDomains = ['example.com', 'mysite.com'];
    const parsedUrl = new URL(url, 'http://localhost');

    if (allowedDomains.includes(parsedUrl.hostname)) {
        res.redirect(url);
    } else {
        res.redirect('/');
    }
});

/**
 * Secure cookie handling
 */
app.get('/secure-cookie', (req, res) => {
    const userId = req.query.userId || '';

    // SAFE: HttpOnly and Secure flags
    res.cookie('sessionId', userId, {
        maxAge: 3600000,
        httpOnly: true,
        secure: process.env.NODE_ENV === 'production',
        sameSite: 'strict'
    });
    res.send('Secure cookie set');
});

/**
 * Safe JSON parsing
 */
function safeJsonParse(jsonString) {
    // SAFE: Using JSON.parse with try-catch
    try {
        return JSON.parse(jsonString);
    } catch (e) {
        throw new Error('Invalid JSON');
    }
}

module.exports = {
    app,
    displayComment,
    domXSS,
    mergeObjects,
    vulnerableRegex,
    unsafeDeserialization,
    escapeHtml,
    safeJsonParse
};

if (require.main === module) {
    app.listen(3000, () => {
        console.log('Server running on port 3000');
    });
}